### PR TITLE
feat: GraphQL upsert mutation

### LINKED ISSUE #783: Support `upsert` mutation
Support `upsert` mutation operation.

### PR DESCRIPTION
**Pull request by nasdf**
## Relevant issue(s)

Resolves #783

## Description

This PR adds a new upsert GraphQL mutation operation.

## Tasks

- [x] I made sure the code is well commented, particularly hard-to-understand areas.
- [x] I made sure the repository-held documentation is changed accordingly.
- [x] I made sure the pull request title adheres to the conventional commit style (the subset used in the project can be found in [tools/configs/chglog/config.yml](tools/configs/chglog/config.yml)).
- [x] I made sure to discuss its limitations such as threats to validity, vulnerability to mistake and misuse, robustness to invalidation of assumptions, resource requirements, ...

## How has this been tested?

Added integration tests.

Specify the platform(s) on which this was tested:
- MacOS

### CODE REVIEW

File: internal/planner/upsert.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+func (n *upsertNode) Source() planNode {
+	return n.results
+}
+
+func (n *upsertNode) simpleExplain() (map[string]any, error) {
+	simpleExplainMap := map[string]any{}
+
+	// Add the filter attribute if it exists, otherwise have it nil.
+	if n.filter == nil {
+		simpleExplainMap[filterLabel] = nil
```
Comment (shahzadlone) [inline internal/planner/upsert.go:130]: question: is it legal to have no filter?

todo: regardless might be an easy test to add without a filter in an explain test to hit this line

**Review by shahzadlone — COMMENTED** [review]
Wowza, super stuff! Left some comments in first past :)

File: tests/integration/mutation/upsert/simple_test.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+				}`,
+				ExpectedError: `cannot upsert multiple matching documents`,
+			},
+		},
+	}
+
+	testUtils.ExecuteTestCase(t, test)
+}
+
+func TestMutationUpsertSimple_WithUniqueCompositeIndexAndDuplicateUpdate_ReturnsError(t *testing.T) {
```
Comment (islamaliev) [inline tests/integration/mutation/upsert/simple_test.go:150]: praise: thanks for this test and for making clear test names.

File: tests/integration/mutation/upsert/simple_test.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+				}`,
+				Results: map[string]any{
+					"upsert_Users": []map[string]any{
+						{
+							"name": "Bob",
+							"age":  int64(40),
+						},
+					},
+				},
+			},
```
Comment (islamaliev) [inline tests/integration/mutation/upsert/simple_test.go:56]: suggestion: I think it also makes sense to run a follow up query request to make it really fetches them from storage and not just as a response.

File: client/request/consts.go [hunk 4]
```diff
@@ -21,6 +21,8 @@ const (
 
 	Cid         = "cid"
 	Input       = "input"
+	CreateInput = "create"
+	UpdateInput = "update"
```
Comment (islamaliev) [inline client/request/consts.go:25]: question: shouldn't "create" and "update" replace "input"?

File: internal/planner/upsert.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+		switch next {
+		case true:
+			n.currentValue = n.results.Value()
+			// make sure multiple documents do not match
+			next, err := n.results.Next()
+			if err != nil {
+				return false, err
+			}
+			if next {
+				return false, ErrUpsertMultipleDocuments
```
Comment (islamaliev) [inline internal/planner/upsert.go:51]: question: is there an inherent problemt with upserting multiple docs?

**Review by islamaliev — COMMENTED** [review]
Looks nice for far! Very useful feature. I have most suggestions.

File: internal/planner/upsert.go [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+// Next only returns once.
+func (n *upsertNode) Next() (bool, error) {
+	if n.isUpserting {
+		next, err := n.results.Next()
+		if err != nil {
+			return false, err
+		}
+
+		switch next {
+		case true:
```
[the remark this answers is not in the document]
Comment (shahzadlone) [reply internal/planner/upsert.go:43]: lol +1 for making into `if`

File: client/request/consts.go (hunk 4 above)
Comment (nasdf) [reply client/request/consts.go:25]: I think keeping the `input` label for create and update mutations is fine.

Changing it to `create_User(create: {name: "Bob"})` doesn't read as nice imo.

File: internal/planner/upsert.go (hunk 5 above)
Comment (nasdf) [reply internal/planner/upsert.go:51]: There's nothing stopping us from supporting it in the future. It's there to stop users from doing something unexpected like updating all documents when they just wanted to upsert a single document.

File: internal/planner/upsert.go [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+	filter      *mapper.Filter
+	createInput map[string]any
+	updateInput map[string]any
+	doc         *client.Document
+	isUpserting bool
+	results     planNode
+}
+
+// Next only returns once.
+func (n *upsertNode) Next() (bool, error) {
```
[the remark this answers is not in the document]
Comment (nasdf) [reply internal/planner/upsert.go:35]: I did have that thought, but I think its worth the extra LOC to keep it separate incase we decide to extend or modify it.

File: tests/integration/mutation/upsert/simple_test.go [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+			testUtils.CreateDoc{
+				Doc: `{
+					"name": "Bob",
+					"age": 50
+				}`,
+			},
+			testUtils.Request{
+				Request: `mutation {
+					upsert_Users(
```
[the remark this answers is not in the document]
Comment (nasdf) [reply tests/integration/mutation/upsert/simple_test.go:315]: I'm on the fence with this. I'd assume most tests are either creating or updating docs explicitly and won't need the added benefits of upsert.

File: internal/request/graphql/schema/descriptions.go [hunk 9]
```diff
@@ -104,6 +104,13 @@ An optional value that specifies as to whether deleted documents may be
 	createDocumentDescription string = `
 Creates one or more documents of this type using the data provided.
 `
+	upsertDocumentDescription string = `
+Update or create a document in this collection using the data provided. The unique
+ document matching the provided filter will be updated with the provided update input,
+ or if no matching document is found, a new document will be created with the
+ provided create input.
+
+NOTE: It is highly recommended to create an index on the fields used to filter.`
```
[the remark this answers is not in the document]
Comment (nasdf) [reply internal/request/graphql/schema/descriptions.go]: done

File: tests/integration/mutation/upsert/simple_test.go [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+						filter: {},
+						create: {name: "Alice", age: 40},
+						update: {age: 50}
+					) {
+						name
+						age
+					}
+				}`,
+				ExpectedError: `cannot upsert multiple matching documents`,
```
[the remark this answers is not in the document]
Comment (nasdf) [reply tests/integration/mutation/upsert/simple_test.go:188]: done

File: tests/integration/mutation/upsert/simple_test.go [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+					upsert_Users(
+						filter: {name: {_eq: "Bob"}},
+						create: {name: "Alice", age: 40},
+						update: {name: "Alice", age: 40}
+					) {
+						name
+						age
+					}
+				}`,
+				ExpectedError: `can not index a doc's field(s) that violates unique index`,
```
[the remark this answers is not in the document]
Comment (nasdf) [reply tests/integration/mutation/upsert/simple_test.go:324]: I think this should be covered in the index docs and not the mutation operation docs.

File: tests/integration/mutation/upsert/simple_test.go [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+			testUtils.CreateDoc{
+				Doc: `{
+					"name": "Alice",
+					"age": 40
+				}`,
+			},
+			testUtils.Request{
+				Request: `mutation {
+					upsert_Users(
+						filter: {},
```
[the remark this answers is not in the document]
Comment (nasdf) [reply tests/integration/mutation/upsert/simple_test.go:180]: added tests for each

File: internal/planner/upsert.go (hunk 1 above)
Comment (nasdf) [reply internal/planner/upsert.go]: A filter is required. I've removed the nil check.

File: internal/planner/upsert.go [hunk 13]
```diff
@@ -0,0 +1,185 @@
+// Copyright 2022 Democratized Data Foundation
```
[the remark this answers is not in the document]
Comment (nasdf) [reply internal/planner/upsert.go]: done

File: internal/planner/upsert.go [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+	docMapper
+
+	p           *Planner
+	collection  client.Collection
+	filter      *mapper.Filter
+	createInput map[string]any
+	updateInput map[string]any
+	doc         *client.Document
+	isUpserting bool
+	results     planNode
```
[the remark this answers is not in the document]
Comment (nasdf) [reply internal/planner/upsert.go]: changed to source

File: internal/planner/upsert.go [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+	documentIterator
+	docMapper
+
+	p           *Planner
+	collection  client.Collection
+	filter      *mapper.Filter
+	createInput map[string]any
+	updateInput map[string]any
+	doc         *client.Document
+	isUpserting bool
```
[the remark this answers is not in the document]
Comment (nasdf) [reply internal/planner/upsert.go]: I copied it from the update node which has an `isUpdating` prop.

File: internal/planner/upsert.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (nasdf) [reply internal/planner/upsert.go]: done

File: internal/planner/upsert.go [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+func (n *upsertNode) Init() error {
+	return n.results.Init()
+}
+
+func (n *upsertNode) Start() error {
+	doc, err := client.NewDocFromMap(n.createInput, n.collection.Definition())
+	if err != nil {
+		return err
+	}
+	n.doc = doc
```
[the remark this answers is not in the document]
Comment (nasdf) [reply internal/planner/upsert.go]: done

File: tests/integration/mutation/upsert/simple_test.go (hunk 3 above)
Comment (nasdf) [reply tests/integration/mutation/upsert/simple_test.go:56]: done

**Review by islamaliev — APPROVED** [review]
Great job, Keenan!

**Review by fredcarle — APPROVED** [review]

**Review by fredcarle — APPROVED** [review]
LGTM!

**Review by shahzadlone — APPROVED** [review]
Thanks for doing the requested changes and answering all the comments =)

**Pull request merged**